pkg/grpc/exec/exec.go

```diff
@@ -415,17 +510,32 @@ func GetProcessExit(event *MsgExitEventUnix) *tetragon.ProcessExit {
 
 type MsgExitEventUnix struct {
 	tetragonAPI.MsgExitEvent
-	RefCntDone [2]bool
+	RefCntDone [3]bool
```

Comment:
Not sure that having one more bool for all ancestors is a correct approach. We use index 0 for process and index 1 for parent and the reason is that we don't want to update refcnt multiple times when calling RetryInternal. 

In the exit event for example, you could possibly `refDec` some of the ancestors and not all. So in the second `RetryInternal` call we will not know which of those we need to `refDec`. I may miss something in that case but worth thinking of that. A more reasonable approach for me could be to have one `RefCntDone` for each ancestor.

This applies also in the other cases where you add one more `RefCntDone` entry.

Reply:
Currently, i call `RefDec` for all ancestors only if i was able to collect them all.
```go
if proc != nil && proc.NeededAncestors() && !msg.RefCntDone[AncestorsRefCnt] {
	if ancestors, perr := process.GetAncestorProcessesInternal(proc.UnsafeGetProcess().ParentExecId); perr == nil {
		var tetragonAncestors []*tetragon.Process
		for _, ancestor := range ancestors {
			tetragonAncestors = append(tetragonAncestors, ancestor.UnsafeGetProcess())
			ancestor.RefDec("ancestor")
		}
		ev.SetAncestors(tetragonAncestors)
		msg.RefCntDone[AncestorsRefCnt] = true
	} else {
		eventcache.CacheRetries(eventcache.AncestorsInfo).Inc()
		err = eventcache.ErrFailedToGetAncestorsInfo
	}
}
```
If `GetAncestorProcessesInternal` returns an error there, i do not decrement ancestors' `refcnt` at all.

Otherwise, if `GetAncestorProcessesInternal` was successful, `msg.RefCntDone[AncestorsRefCnt]` will be set to `true` and in the next iteration upper condition will result in `false`. So another `RefDec` for ancestors will not be called.

I thought of trying to decrement `refcnt` of ancestors even if i was not able to collect them all, but in that case in the next `RetryInternal` iteration i can find myself in a situation where i can't get any ancestors anymore. As i have already decremented some of their `refcnt` before and some of them could've ended up with their `refcnt` set to 0. So, an alternative can be to store an `ExecID` of the last ancestor we were able to find in the current iteration and to start collection from that in the next iteration.

I guess, i can try doing something like this then:
```go
if msg.AncestorExecId != "" {
	execId = msg.AncestorExecId
} else if proc != nil {
	execId = proc.UnsafeGetProcess().ParentExecId
}

if execId != "" && proc != nil && proc.NeededAncestors() && !msg.RefCntDone[AncestorsRefCnt] {
	tetragonAncestors := ev.GetAncestors()
	ancestors, perr := process.GetAncestorProcessesInternal(execId)
	for _, ancestor := range ancestors {
		tetragonAncestors = append(tetragonAncestors, ancestor.UnsafeGetProcess())
		ancestor.RefDec("ancestor")
	}
	ev.SetAncestors(tetragonAncestors)
	if perr == nil {
		msg.RefCntDone[AncestorsRefCnt] = true
	} else {
		if len(ancestors) > 0 {
			msg.AncestorExecId = ancestors[len(ancestors)-1].UnsafeGetProcess().ExecId
		}
		eventcache.CacheRetries(eventcache.AncestorsInfo).Inc()
		err = eventcache.ErrFailedToGetAncestorsInfo
	}
}
```
Will that be better?